src/Pipeline.cpp

```diff
@@ -673,109 +675,99 @@ struct JITFuncCallContext {
-            OutputBufferType obt = {f, t, f.dimensions()};
-            output_buffer_types.push_back(obt);
-        }
-    }
-
-    user_assert(output_buffer_types.size() == dst.size())
-        << "Realization contains wrong number of Images (" << dst.size()
-        << ") for realizing pipeline with " << output_buffer_types.size()
-        << " outputs\n";
-
-    // Check the type and dimensionality of the buffer
-    for (size_t i = 0; i < dst.size(); i++) {
-        Function func = output_buffer_types[i].func;
-        int  dims = output_buffer_types[i].dims;
-        Type type = output_buffer_types[i].type;
-        user_assert(dst[i].dimensions() == dims)
-            << "Can't realize Func \"" << func.name()
-            << "\" into Buffer at " << (void *)dst[i].data()
-            << " because Buffer is " << dst[i].dimensions()
-            << "-dimensional, but Func \"" << func.name()
-            << "\" is " << dims << "-dimensional.\n";
-        // For our purposes here, consider all Handle types equivalent:
-        // Buffer<> doesn't retain handle-traits (thus it collapses all
-        // all Handle types into void*), but Func does not, so we can have
-        // confusing cases where Buffer<char*> is not "compatible" with Func<char*>.
-        // (Buffer-of-handle-type is a degenerate case anyway...)
-        user_assert(dst[i].type() == type ||
-                    (dst[i].type().is_handle() && type.is_handle()))
-            << "Can't realize Func \"" << func.name()
-            << "\" into Buffer at " << (void *)dst[i].data()
-            << " because Buffer has type " << Type(dst[i].type())
-            << ", but Func \"" << func.name()
-            << "\" has type " << type << ".\n";
-    }
-
+    const bool no_param_map = &param_map == &ParamMap::empty_map();
 
     // Come up with the void * arguments to pass to the argv function
-    vector<const void *> arg_values;
-
+    size_t arg_index = 0;
     for (const InferredArgument &arg : contents->inferred_args) {
         if (arg.param.defined()) {
             if (arg.param.same_as(contents->user_context_arg.param)) {
-                arg_values.push_back(user_context);
+                args_result.store[arg_index++] = user_context;
             } else {
-                Buffer<> *buf_out_param;
-                const Parameter &p = param_map.map(arg.param, buf_out_param);
-                if (!is_bounds_inference) {
-                    user_assert(buf_out_param == nullptr) << "Cannot pass Buffer<> pointers in parameters map to a compute call.\n";
-                }
+                Buffer<> *buf_out_param = nullptr;
+                const Parameter &p = no_param_map ? arg.param : param_map.map(arg.param, buf_out_param);
+                user_assert(is_bounds_inference || !buf_out_param)
+                    << "Cannot pass Buffer<> pointers in parameters map to a compute call.\n";
 
                 if (p.is_buffer()) {
                     // ImageParam arg
```

Comment:
Could this just be:
```
if (p.is_buffer()) {
  args_result.store[arg_index++] = p.raw_buffer();
} else {
  args_result.store[arg_index++] = p.scalar_address();
}
```

Reply:
The `infer_input_bounds` case passes in Buffers which are not defined. I'm pretty sure `Parameter::raw_buffer` will signal an error in that case rather than returning `nullptr`.